luarules/gadgets/cmd_nano_range_check.lua

```diff
@@ -0,0 +1,160 @@
+function gadget:GetInfo()
+	return {
+		name = "Construction Turrets Range Check",
+		desc = "Stops construction turrets from getting stuck on orders out of reach.",
+		author = "Nehroz",
+		date = "2024.12.01",
+		layer = 0,
+		enabled = true,
+	}
+end
+
+if not gadgetHandler:IsSyncedCode() then return end
+
+local CMD_STOP = CMD.STOP
+local gMaxUnits = Game.maxUnits
+local updateInterval = Game.gameSpeed / 2
+
+local spValidUnitID = Spring.ValidUnitID
+local spValidFeatureID = Spring.ValidFeatureID
+local spGetUnitDefID = Spring.GetUnitDefID
+local spGetUnitSeparation = Spring.GetUnitSeparation
+local spGetUnitFeatureSeparation = Spring.GetUnitFeatureSeparation
+local spGetUnitCommands = Spring.GetUnitCommands
+local spGiveOrderToUnit = Spring.GiveOrderToUnit
+
+--- Key is the nano's ID, followed by a set-like with the unitID of target currently inside the nano's range.
+--- Used for CommandArray manipulation
+---@type {[integer] : {integer : {cmdTag : integer, status : integer}}}
+local trackingTable = {}
+-- Limits how many GetUnitCommands are called inside a frame when required.
+-- higher values allow faster tagging synchronization after a new command;
+-- which can cause proportional lag when many units are targeted. (long tables returned)
+local tagUpdateFrameMax = 1
+
+local statuses = {
+	outOfRange = 0,
+	inRange = 1,
+}
+
+local mobileUnits = {}
+local constructionTurretsDefs = {}
+
+for unitDefID, unitDef in ipairs(UnitDefs) do
+	if unitDef.isFactory == false and unitDef.isStaticBuilder then
+		constructionTurretsDefs[unitDefID] = {
+			maxBuildDistance = unitDef.buildDistance + unitDef.radius,
+			buildRange3D = unitDef.buildRange3D,
```

Comment:
here likely could avoid the table and just use array so accessing will be cheaper

Reply:
There is nothing more annoying or frustrating when developing the code base than having to deal with `someFunc(someTable[3], someTable[2])`. It is a multiplier for development time.

If the performance penalties are too much to bear, then we'll have to develop a post-processor to "compile" our Lua code to replace hashed tables with arrays.